replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/ReplicatorRunner.java

```diff
@@ -0,0 +1,115 @@
+import org.codice.ditto.replication.api.ReplicatorConfigLoader;
+import org.codice.ditto.replication.api.SyncRequest;
+import org.codice.ditto.replication.api.impl.data.SyncRequestImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The ReplicatorRunner periodical queues up replication jobs for all the current replication
+ * configurations.
+ */
+public class ReplicatorRunner {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ReplicatorRunner.class);
+
+  private final Security security;
+
+  private final Replicator replicator;
+
+  private final ReplicatorConfigLoader replicatorConfigLoader;
+
+  private final ScheduledExecutorService scheduledExecutor;
+
+  public ReplicatorRunner(
+      ScheduledExecutorService scheduledExecutor,
+      Replicator replicator,
+      ReplicatorConfigLoader replicatorConfigLoader) {
+    this(scheduledExecutor, replicator, replicatorConfigLoader, Security.getInstance());
+  }
+
+  public ReplicatorRunner(
+      ScheduledExecutorService scheduledExecutor,
+      Replicator replicator,
+      ReplicatorConfigLoader replicatorConfigLoader,
+      Security security) {
+    this.scheduledExecutor = notNull(scheduledExecutor);
+    this.replicator = notNull(replicator);
+    this.replicatorConfigLoader = notNull(replicatorConfigLoader);
+    this.security = security;
+  }
+
+  public void init() {
+    long period =
+        Long.parseLong(
+            System.getProperty(
+                "org.codice.replication.period", String.valueOf(TimeUnit.MINUTES.toSeconds(5))));
+    scheduledExecutor.scheduleAtFixedRate(
+        this::replicateAsSystemUser, TimeUnit.MINUTES.toSeconds(1), period, TimeUnit.SECONDS);
+    LOGGER.info("Replication checks scheduled for every {} seconds.", period);
+  }
+
+  public void destroy() {
+    scheduledExecutor.shutdownNow();
+  }
+
+  @VisibleForTesting
+  void replicateAsSystemUser() {
+    security.runAsAdmin(
+        () -> {
+          try {
+            security.runWithSubjectOrElevate(
+                () -> {
+                  scheduleReplication();
+                  return null;
+                });
+          } catch (SecurityServiceException | InvocationTargetException e) {
+            LOGGER.debug("Error scheduling replication.", e);
+          }
+          return null;
+        });
+  }
+
+  @VisibleForTesting
+  void scheduleReplication() {
+    replicatorConfigLoader
+        .getAllConfigs()
+        .stream()
+        .map(config -> new SyncRequestImpl(config, new ReplicationStatus(config.getName())))
+        .forEach(this::submitSyncRequest);
+  }
+
+  private void submitSyncRequest(SyncRequest request) {
+    try {
+      LOGGER.debug("Submitting replication request for {} ", request.getConfig().getName());
+      replicator.submitSyncRequest(request);
+    } catch (InterruptedException e) {
+      LOGGER.debug("Exception received from the replicator. ", e);
+      Thread.currentThread().interrupt();
```

Comment:
❗️ The problem in doing this here is that you are in the middle of an iteration controlled by `scheduleReplication()` which means that the next iteration will be interrupted again. Typically, at a low level like this it is better to throw back an `InterruptedException` and let the code where the loop is handle it properly. Since in your case you want to abort the loop (I assume). The calling code would catch the exception outside of the loop and propagate back the interruption state on the thread before bailing out.

Reply:
Reworked calling method to fix this